### ISSUE TITLE
Use Client.as_current more

### ISSUE THREAD

**Issue by mrocklin** [dask/distributed#7377]
I've been playing around with a workflow that uses many Clients in many threads.  Each client looks at a remote Dask cluster running a large XGBoost computation.  I run them all in the same process to use Optuna for HPO.

I find that when run naively I run into issues with different threads picking up the wrong client at times.  There are various parts in the codebase where we call `get_client` or `default_client` and it's not always foolproof which one gets selected.

My approach to this was to use `Client.current()` everywhere in `default_client` and `get_client()` and then to wrap my code in a `client.as_current` context manager

```python
def f():
    with Cluster(...) as cluster:
        with Client(cluster) as client:
            with client.as_current:
                XGBoost stuff

with concurrent.futures.ThreadPoolExecutor() as e:
    futures = [e.submit(f) for _ in range(10)]
    results = [future.result() for future in futures]
```

We have too many ways to get "the client" clearly.  It seems like some attention was paid to using `ContextVars` with `Client.current`.  This seems like a sensible approach to me and seems to be decently well done.  We might now want to leverage it more fully.

For example, there's this in dask/dask

```diff
diff --git a/dask/base.py b/dask/base.py
index 43b0b1adf..bb7ac9d3d 100644
--- a/dask/base.py
+++ b/dask/base.py
@@ -1370,9 +1370,12 @@ def get_scheduler(get=None, scheduler=None, collections=None, cls=None):
                     )
                 return named_schedulers[scheduler]
             elif scheduler in ("dask.distributed", "distributed"):
-                from distributed.worker import get_client
+                from dask.distributed import default_client, get_client

-                return get_client().get
+                try:
+                    return default_client().get
+                except ValueError:
+                    return get_client().get
             else:
                 raise ValueError(
                     "Expected one of [distributed, %s]"
```

Also, maybe `Client.__enter__` and `Client.__aenter__` should also call `client.as_current.__enter__`?

**Comment by fjetter** [comment]
I started on something similar a couple of months ago but never pushed it over the finishing line, see https://github.com/dask/distributed/pull/6527

My gut feeling is that we should get rid of the overly complex default client mechanism with a couple of targeted "as_current" calls.

For some problems with the current default client, see also https://github.com/dask/distributed/pull/5901 (I never merged that one since I didn't feel comfortable with the complexity)

**Comment by mrocklin** [comment]
Concretely short term I mostly need https://github.com/dask/dask/pull/9723 and https://github.com/dask/distributed/pull/7369 in.  Is this feasible, or do these cause the issues that you're referring to above?

**Comment by fjetter** [comment]
There are a couple of different thread safety related PRs that may need to be merged before this

I think https://github.com/dask/distributed/pull/5901 (should be ready to merge) will be helpful for the xgboost issue since we're relying on the default client mechanism

There are two more issues that affect `worker.get_client` but these should be bypassed when relying on `Client.current` IIUC
- https://github.com/dask/distributed/pull/5467 (almost done, suggests that switching to as_current may resolve the issue)
- related issues https://github.com/dask/distributed/issues/5466 / https://github.com/dask/distributed/issues/3827

**Comment by mrocklin** [comment]
Indeed, I just tried the notebook again without this change and things seem ok.  The other PR is necessary though I think.  xgboost.dask calls `distributed.wait` which calls `default_client` which should check `Client.current` I think.  We could also modify `distributed.wait`.  That's the real API boundary that's relevant here.

**Comment by mrocklin** [comment]
Relatedly, I'd also love to see `with Client` invoke `with client.as_current` by default.  Any objection or concerns to this?  

Is this something that your team could get in by any chance?

**Comment by fjetter** [comment]
I think most of what you're looking for is in https://github.com/dask/distributed/pull/6527
Back then, tests passed. I'll rebase to see if something horrible breaks

**Comment by mrocklin** [comment]
My concern looking at comments was that maybe that PR wouldn't get in soon
and maybe this one would.  If that's not the case (in either direction)
then I'm happy to follow your lead here.

**Comment by fjetter** [comment]
https://github.com/dask/distributed/pull/6527 is the context manager change you proposed + a few extras. Not the bigger thread safety thing.
definitely +1 for the ctxmanager + current
